modules/face/src/facemarkLBF.cpp

```diff
@@ -1191,6 +1181,185 @@ namespace face {
         free(Y);
     } // Regressor:globalRegressionTrain
 
+    /*adapted from the liblinear library*/
+    /* TODO: change feature_node to MAT
+    * as the index in feature_node is only used for "counter"
+    */
+    Mat FacemarkLBFImpl::Regressor::supportVectorRegression(
+        feature_node **x, double *y, int nsamples, int feat_size, bool verbose
+    ){
+        #define GETI(i) ((int) y[i])
+
+        std::vector<double> w;
+        w.resize(feat_size);
+
+        int l = nsamples; // n-samples
+        double C = 1./(double)nsamples;
+        double p = 0;
+        int w_size = feat_size; //feat size
+        double eps =  0.00001;
+        int i, s, iter = 0;
+        int max_iter = 1000;
+        int active_size = l;
+        std::vector<int> index(l);
+
+        double d, G, H;
+        double Gmax_old = HUGE_VAL;
+        double Gmax_new, Gnorm1_new;
+        double Gnorm1_init = -1.0; // Gnorm1_init is initialized at the first iteration
+        std::vector<double> beta(l);
+        std::vector<double> QD(l);
+
+        double lambda[1], upper_bound[1];
+        lambda[0] = 0.5/C;
+        upper_bound[0] = HUGE_VAL;
+
+        // Initial beta can be set here. Note that
+        // -upper_bound <= beta[i] <= upper_bound
+        for(i=0; i<l; i++)
+            beta[i] = 0;
+
+        for(i=0; i<w_size; i++)
+            w[i] = 0;
+
+        for(i=0; i<l; i++){
+            QD[i] = 0;
+            feature_node *xi = x[i];
+            while(xi->index != -1){
+                double val = xi->value;
+                QD[i] += val*val;
+                w[xi->index-1] += beta[i]*val;
+                xi++;
+            }
+            index[i] = i;
+        }
+
+        while(iter < max_iter){
+            Gmax_new = 0;
+            Gnorm1_new = 0;
+
+            for(i=0; i<active_size; i++){
+                int j = i+rand()%(active_size-i);
```

Comment:
Please avoid using of `rand()` - it doesn't provide reproducible results.

Like "stereo" test (it uses `rand()` too), which fails time-to-time.